### PR TITLE
MINOR: Improve `datafusion-proto` API

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

N/A

 # Rationale for this change

New unified API for serializing plans and expressions.

# What changes are included in this PR?

``` rust
let serializer = Serializer::new()
  .with_function_registry(...)
  .with_extension_codec(...);

let bytes = serializer.serialize_plan(...);

let bytes = serializer.serialize_expr(...);
```

# Are there any user-facing changes?

API change

# Does this PR break compatibility with Ballista?

TBD

### CODE REVIEW

File: datafusion/proto/src/lib.rs [hunk 1]
```diff
@@ -33,6 +42,122 @@ pub mod to_proto;
...
+            _ => Serializeable::from_bytes(bytes),
+        }
+    }
+
+    /// Serialize a logical plan
+    pub fn serialize_plan(&self, plan: &LogicalPlan) -> Result<Vec<u8>> {
+        //TODO note that we will need to also pass the function_registry along once we
+        // add support for subquery expressions in https://github.com/apache/arrow-datafusion/issues/2640
+        let protobuf = match self.extension_codec {
+            Some(codec) => protobuf::LogicalPlanNode::try_from_logical_plan(plan, codec)?,
```
Comment (tustvold) [inline datafusion/proto/src/lib.rs:105]: Is your plan to move this logic onto Serializer or to pass function_registry down?

**Review by tustvold — APPROVED** [review]
I'm not very familiar with this code, and so you possibly want to wait for approval from someone else, but I think from a pure code standpoint this is fine.

However, I do notice that the current code is pretty consistently functional and this is effectively grafting an OOP abstraction on top of it. This isn't an inherently bad thing, but I wonder if you've given thought to `impl Serializable for LogicalPlan` or something similar?

FWIW I suspect this approach would make recursion into subqueries easier as well.

**Comment by alamb** [comment]
marking as draft as there is future work planned here

**Comment by andygrove** [comment]
I am closing this since we have a working solution over in Ballista now (https://github.com/apache/arrow-ballista/pull/57) and I will let someone else pick up this task of improving the API here. I can try again later if nobody does pick this up.

**Pull request closed without merging**